src/store/index.ts

```diff
@@ -93,7 +94,8 @@ export const storeConfig: StoreOptions<State> = {
     notifications,
     onboarding,
   },
-  state: initialState,
+  // Note: *Technically*, `initialState` here doesn’t hold all properties from `State`. All module state is missing. However, the module state MUST NOT be made optional in `State` as otherwise, it will be considered as optional when accessing, for example, `store.state.config.status`.
```

Comment:
This going to be a longer explanation, so strap in. (Note: in the following, I’m calling `RootState` what in our app is simply called `State` to distinguish things)

For a store like ours, there is a `RootState` which is the type of all state that belongs to the store created with `createStore` in `@/store/store.ts`. This store _also_ has **namespaced** modules which introduce more state under named keys. Each module state also has a type (e.g. `ModuleState`).

With this, we have global state (e.g. `state.key`) and module state `state.moduleName.key`. The way Vuex sets up its types imply that `RootState` does not include module state. Taking this further: creating a store receives an initial store configuration (with the type of `RootState`. This operation returns a `store` object with a `state` property which _also_ has the type `RootState`. With what I explained above, this is not correct. the resulting store’s type _does_ include the module state (under the respective module names), but it doesn’t seem to be possible on Vuex’s side to correctly infer this type relation.

The way how I initialize our store is a compromise of this problem that ensures that we can access `store.state.moduleName.key` without the need to _always_ type-check whether `moduleName` exists. We know it does because we always add it.

I asked about a better way for this in the Vue discord where I provided the following minimal example illustrating the problem. It includes the two steps I also used in Kuma GUI to mitigate most of the problem (i.e. what you asked about).

---

Here is a minimal example illustrating the problem:

```typescript
import { createStore, Module, StoreOptions } from 'vuex'

const rootState = {
    key: 'value',
}

const testModuleState = {
    moduleKey: 'moduleValue',
}

/**
 * **Issue 1**: Defining the root state like this
 * makes it impossible to access module state with,
 * for example, `store.state.testModule.moduleKey`.
 * The resulting TypeScript error occurs:
 *
 * ~~~
 * store.state.testModule.moduleKey
 * //          ^^^^^^^^^^
 * // Property 'testModule' does not exist on type '{ key: string; }'.
 * ~~~
 *
 * That's because the type assigned to the result of calling `createStore`
 * does in fact not include state from modules.
 *
 * As a workaround, one can add the module state manually.
 */
type RootState = typeof rootState// & { testModule: TestModuleState }

type TestModuleState = typeof testModuleState

const testModule: Module<TestModuleState, RootState> = {
    namespaced: true,
    state: testModuleState,
}

const storeOptions: StoreOptions<RootState> = {
    /**
     * **Issue 2**: When adding, via intersection as shown above,
     * the module state to `RootState`,
     * the following TypeScript error occurs:
     *
     * ~~~
     *    state: rootState
     * // ^^^^^
     * // Type '{ key: string; }' is not assignable to type 'RootState | (() => RootState)'.
     * ~~~
     *
     * As a workaround, one can manually cast `state as RootState`,
     * but this loses you true TypeScript errors.
     */
    state: rootState/* as RootState */,
    modules: {
        testModule,
    },
}

const store = createStore<RootState>(storeOptions)

store.state.key
store.state.testModule.moduleKey
//          ^^^^^^^^^^
// Property 'testModule' does not exist on type '{ key: string; }'. ts(2339)
```

Reply:
And to actually directly answer the following question:

> why is it a problem that it's optional?

We would need to always type-guard access to module state which is **always** unnecessary because we know it exists as long as our state types are complete and accurate.

`store.state.config.status` always exists once we created a store, but we would need to at least white `store.state.config?.status` to read it and use an if statement to write to it (because you can’t use optional chaining on the left hand side of an assignment).

Reply:
jeez, this is a blogpost on it's own 😄 thanks for explaining!